src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,387 @@
+                if ( encodingPreference == FileEncoding.WindowsLegacy )
+                {
+                    resolvedEncoding = GetWindowsLegacyEncoding(name);
+                }
+                else if ( encodingPreference != FileEncoding.Unknown )
+                {
+                    resolvedEncoding = GetEncoding(encodingPreference);
+                }
+                // the final else would be set the encoding to GetDefaultEncoding() which was handled above
+            }
+
+            return resolvedEncoding;
+        }
+
+        // [System.Text.Encoding]::GetEncodings() | ? { $_.GetEncoding().GetPreamble() } |
+        //     Add-Member ScriptProperty Preamble { $this.GetEncoding().GetPreamble() -join "-" } -PassThru |
+        //     Format-Table -Auto
+        internal static Dictionary<String, FileEncoding> encodingMap =
+            new Dictionary<string, FileEncoding>()
+            {
+                { "255-254", FileEncoding.Unicode },
+                { "254-255", FileEncoding.BigEndianUnicode },
+                { "255-254-0-0", FileEncoding.UTF32 },
+                { "0-0-254-255", FileEncoding.BigEndianUTF32 },
+                { "239-187-191", FileEncoding.UTF8 },
+            };
+
+        internal static char[] nonPrintableCharacters = {
+            (char) 0, (char) 1, (char) 2, (char) 3, (char) 4, (char) 5, (char) 6, (char) 7, (char) 8,
+            (char) 11, (char) 12, (char) 14, (char) 15, (char) 16, (char) 17, (char) 18, (char) 19, (char) 20,
+            (char) 21, (char) 22, (char) 23, (char) 24, (char) 25, (char) 26, (char) 28, (char) 29, (char) 30,
+            (char) 31, (char) 127, (char) 129, (char) 141, (char) 143, (char) 144, (char) 157 };
+
+        internal static readonly UTF8Encoding utf8NoBom = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+        // take a look at the file contents and guess at the best encoding
+        internal static FileEncoding GetEncoding(string path)
+        {
+            if (!File.Exists(path))
+            {
+                return FileEncoding.Default;
+            }
+
+            byte[] initialBytes = new byte[100];
+            int bytesRead = 0;
+
+            try
+            {
+                using (FileStream stream = System.IO.File.OpenRead(path))
+                {
+                    using (BinaryReader reader = new BinaryReader(stream))
+                    {
+                        bytesRead = reader.Read(initialBytes, 0, 100);
+                    }
+                }
+            }
+            catch (IOException)
+            {
+                return FileEncoding.Default;
+            }
+
+            // Test for four-byte preambles
+            string preamble = null;
+            FileEncoding foundEncoding = FileEncoding.Default;
+
+            if (bytesRead > 3)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1], initialBytes[2], initialBytes[3]);
```

Comment:
this was code which was refactored from another location, I didn't really want to change it